src/raglite/_database.py

```diff
@@ -304,31 +295,138 @@ def create_database_engine(config: RAGLiteConfig | None = None) -> Engine:
                     INSERT INTO keyword_search_chunk_index(rowid, body) VALUES (new.rowid, new.body);
                 END;
-                """)
+                """
+                )
             )
             session.commit()
     return engine
+
+
+@dataclass
+class ContextSegment:
+    """A class representing a segment of context from a document.
+
+    This class holds information about a specific segment of a document,
+    including its document ID and associated chunks of text with their IDs and scores.
+
+    Attributes
+    ----------
+        document_id (str): The unique identifier for the document.
+        chunks (list[Chunk]): List of chunks for this segment.
+        chunk_scores (list[float]): List of scores for each chunk.
+
+    Raises
+    ------
+        ValueError: If document_id is empty or if chunks is empty.
+    """
+
+    document_id: str
+    chunks: list[Chunk]
+    chunk_scores: list[float]
+
+    def __post_init__(self) -> None:
+        """Validate the segment data after initialization."""
+        if not isinstance(self.document_id, str) or not self.document_id.strip():
+            msg = "document_id must be a non-empty string"
+            raise ValueError(msg)
+        if not self.chunks:
+            msg = "chunks cannot be empty"
+            raise ValueError(msg)
+        if not all(isinstance(chunk, Chunk) for chunk in self.chunks):
+            msg = "all elements in chunks must be Chunk instances"
+            raise ValueError(msg)
+
+    def to_xml(self, indent: int = 4) -> str:
+        """Convert the segment to an XML string representation.
+
+        Args:
+            indent (int): Number of spaces to use for indentation.
+
+        Returns
+        -------
+            str: XML representation of the segment.
+        """
+        chunks_content = "\n".join(str(chunk) for chunk in self.chunks)
+
+        # Create the final XML
+        chunk_ids = ",".join(self.chunk_ids)
+        xml = f"""<document id="{escape(self.document_id)}" chunk_ids="{escape(chunk_ids)}">\n{escape(str(chunks_content))}\n</document>"""
+
+        return xml
+
+    def score(self, scoring_function: Callable[[list[float]], float] = sum) -> float:
+        """Return an aggregated score of the segment, given a scoring function."""
+        return scoring_function(self.chunk_scores)
+
+    @property
+    def chunk_ids(self) -> list[str]:
+        """Return a list of chunk IDs."""
+        return [chunk.id for chunk in self.chunks]
+
+    def __str__(self) -> str:
+        """Return a string representation reconstructing the document with headings.
+
+        Treats headings as a stack, showing headers only when they differ from
+        the current stack path.
+
+        For example:
+        - "# A ## B" shows both headers
+        - "# A ## B" shows nothing (already seen)
+        - "# A ## C" shows only "## C" (new branch)
+        - "# D ## B" shows both (new path)
+        """
```

Comment:
I don't think I understand the purpose of this, could you explain it here?

Reply:
Before this PR, headings were only added to the first chunk and all the other chunk headings were lost as we were concatenating the bodies. In order to solve that we want to keep the headings of all the chunks, but we want to avoid repeating the ones that are already been written. 

This method does the following:
- For each chunk it compares its headings with the current level and writes the ones that are not in the current heading stack.
- It updates the stack to reflect the current heading "state"
- It doesn't check that the heading level is correct (number of '#") as to make it more resilient to weird md conversions. 

This method gives a more accurate reconstruction that the one that it was being used, while avoiding heading repetitions.

Reply:
Actually it was not working as intended. I was assuming that the inner heading was removed from the body, but it is not the case, therefore it's way easier to reconstruct.

```python
    def reconstructed_str(self) -> str:
        """Return a string representation reconstructing the document with headings."""

        heading = self.chunks[0].headings if self.chunks else ""
        bodies = "\n".join([chunk.body for chunk in self.chunks])

        return f"{heading}\n\n{bodies}".strip()
```